src/cascadia/TerminalApp/CascadiaSettingsSerialization.cpp

```diff
@@ -129,6 +134,54 @@ std::unique_ptr<CascadiaSettings> CascadiaSettings::LoadAll()
     // If this throws, the app will catch it and use the default settings
     resultPtr->_ValidateSettings();
 
+    // GH 3855 - Gathering Data on custom profiles to inform better defaults
+    // Do it after everything else so it won't happen unless validation passed.
+    {
+        auto guid = resultPtr->GlobalSettings().GetDefaultProfile();
+
+        // Compare to the defaults.json one that we set on install.
+        // If it's different, log what the user chose.
+        if (hardcodedDefaultGuid != guid)
+        {
+            TraceLoggingWrite(
+                g_hTerminalAppProvider, // handle to TerminalApp tracelogging provider
+                "CustomDefaultProfile",
+                TraceLoggingDescription("Event emitted when user has chosen a different default profile than hardcoded one on load/reload"),
+                TraceLoggingGuid(guid, "DefaultProfile", "ID of user-chosen default profile"),
+                TraceLoggingKeyword(MICROSOFT_KEYWORD_MEASURES),
+                TelemetryPrivacyDataTag(PDT_ProductAndServiceUsage));
+        }
+
+        // If the user had keybinding settings preferences, we want to learn from them to make better defaults
+        auto userKeybindings = resultPtr->_userSettings[JsonKey(KeybindingsKey)];
+        if (!userKeybindings.empty())
+        {

```

Comment:
Hate to do this to you; feel free to shout at me if it's too much work.

Moving this into a function that's called _inside_ the `TraceLoggingWrite` macro will defer all the work, possibly _forever_ if tracing is not actively enabled for this session/user/machine.

This just seems costly to do every time. Thoughts?

Reply:
Nnnnnn as far as I remember, that would count as "doing something cute" with the TraceLoggingMacros. But as far as I remember, there's something in the `conhost.exe` code that lets you check if a session is tracing or not so you can disable yourself from hard-processing work. Let me go see if I can get that here.

Reply:
OK, I found it and put it in.